File: archive/frontend-server-side-recommendation/recommendation/tokenize.ts
const STOP_WORDS = new Set([
  "a",
  "an",
  "and",
  "are",
  "but",
  "for",
  "from",
  "game",
  "games",
  "i",
  "if",
  "in",
  "it",
  "like",
  "of",
  "or",
  "something",
  "that",
  "the",
  "to",
  "want",
  "with",
]);

export function tokenize(text: string) {
  return text
    .toLowerCase()
    .replace(/[^a-z0-9\s-]/g, " ")
    .split(/\s+/)
    .map((token) => token.trim())
    .filter((token) => token.length > 1 && !STOP_WORDS.has(token));
}

export function overlapScore(a: string[], b: string[]) {
  if (!a.length || !b.length) {
    return 0;
  }

  const aSet = new Set(a);
  const bSet = new Set(b);
  let shared = 0;

  for (const token of aSet) {
    if (bSet.has(token)) {
      shared += 1;
    }
  }

  return shared / Math.max(aSet.size, bSet.size);
}

File: backend/app/services/recommendation/tokenize.py
STOP_WORDS = {
    "a",
    "an",
    "and",
    "are",
    "but",
    "for",
    "from",
    "game",
    "games",
    "i",
    "if",
    "in",
    "it",
    "like",
    "of",
    "or",
    "something",
    "that",
    "the",
    "to",
    "want",
    "with",
}


def tokenize(text: str) -> list[str]:
    normalized = "".join(char if char.isalnum() or char in {" ", "-"} else " " for char in text.lower())
    return [
        token
        for token in (part.strip() for part in normalized.split())
        if len(token) > 1 and token not in STOP_WORDS
    ]


def overlap_score(a: list[str], b: list[str]) -> float:
    if not a or not b:
        return 0.0

    a_set = set(a)
    b_set = set(b)
    shared = sum(1 for token in a_set if token in b_set)
    return shared / max(len(a_set), len(b_set))
